Docs: document the rest of the modules (and other cleanups) (#807)
svix/diom: docs/content/modules/rate-limit.mdx

# Rate-limiting (`rate-limit`)

The `rate-limit` module enables developers to limit access to resources such as API endpoints and background workers.

> The module implements the [token bucket algorithm](https://en.wikipedia.org/wiki/Token_bucket) which provides high performance and high flexibility.

Reviewer: ```suggestion
The module implements the [token bucket algorithm](https://en.wikipedia.org/wiki/Token_bucket), which provides high performance and high flexibility.
```

## Configuring the rate limiter

When using the `rate-limit` module you pass the wanted configuration on every call, giving you full flexibility to dynamically configure how you enforce limits.

The configuration lets you configure:
* Capacity: what is the maximum amount of rate-limit tokens this rate-limiter can have.
* Refill rate: how many tokens to refill on every internal.
* Refill interval: how often to refill the amount configured in refill rate (default is one second).

Here is an example configuration:

```rust
// Capacity: 10, refill rate: 1, refill interval: 1 second.
let config = RateLimitTokenBucketConfig::new(10, 1)
    .with_refill_interval(Duration::from_secs(1));
```

## Using the rate limiter (`rate-limit.limit`)

To use the rate limiter you call the `rate-limit.limit` operation with a configuration like so:

```rust
let key = "user:42:api".to_string();
let config = RateLimitTokenBucketConfig::new(10, 1)
    .with_refill_interval(Duration::from_secs(1));

let out = client.rate_limit()

>     .limit(RateLimitCheckIn::new(key.clone(), config.clone()))

Reviewer 2: Do we need the `clone()`s in the examples? they make them look worse
 
```suggestion
    .limit(RateLimitCheckIn::new(key, config))
```

same for the others

Author: Correct rust. I agree it's a bit ugly. :|

## Get remaining tokens (`rate-limit.get-remaining`)

You can use the `rate-limit.get-remaining` function to get the amount of tokens remaining without consuming any.

```rust
// Peek at the remaining tokens without consuming any.
let remaining = client.rate_limit()
    .get_remaining(RateLimitGetRemainingIn::new(key.clone(), config.clone()))
    .await?;
// Retry after tells you when you'll have sufficient tokens available.
let retry_ms = remaining.retry_after.unwrap_or_default().as_millis();
println!("Tokens remaining {}, retry after {retry_ms}ms", remaining.remaining);
```

## Reset the bucket to capacity (`rate-limit.reset`)

To expire a cache value (delete the key), call the `cache.delete` operation with the wanted key:

```rust
// Reset the bucket back to full capacity.
client.rate_limit().reset(RateLimitResetIn::new(key.clone(), config.clone()))
    .await?;
```

## Examples

Here are a few examples of common use-cases when implementing rate-limiting.

### Allow temporary bursts over the limit

A common use-case with rate-limiting is to allow temporary bursts (that go over the allowed rate-limit), but still enforcing the rate-limit on an ongoing basis.

> To achieve it you would want to configure the capacity to the maximum spike capacity, and the refill rate to you wanted limit. For example, if you want to allow for 100 tokens per second with spikes of 200 tokens, you can configure capacity to be 200 and refill rate to 100.

Reviewer: ```suggestion
To achieve it you would want to configure the capacity to the maximum spike capacity, and the refill rate to your desired limit. For example, if you want to allow for 100 tokens per second with spikes of 200 tokens, you can configure capacity to be 200 and refill rate to 100.
```

> This works because it lets users consume 200 tokens, but the refill happens much more slowly, which means that under load they would be limited to 100 tokens per internal, but if the tokens had enough time to replenish to capacity, they would be able to go over the limit.

Reviewer: ```suggestion
This works because it lets users consume 200 tokens, but the refill happens much more slowly, which means that under load they would be limited to 100 tokens per interval, but if the tokens had enough time to replenish to capacity, they would be able to go over the limit.
```

Example configuration:
```rust
RateLimitTokenBucketConfig::new(200, 100)
    .with_refill_interval(Duration::from_secs(1));

> ```

Reviewer 2: this is repeated below

Author: Fixed, thanks!

}

// Request allowed!
```

### Allow temporary bursts over the limit

A common use-case with rate-limiting is to allow temporary bursts (that go over the allowed rate-limit), but still enforcing the rate-limit on an ongoing basis.

To achieve it you would want to configure the capacity to the maximum spike capacity, and the refill rate to you wanted limit. For example, if you want to allow for 100 tokens per second with spikes of 200 tokens, you can configure capacity to be 200 and refill rate to 100.

This works because it lets users consume 200 tokens, but the refill happens much more slowly, which means that under load they would be limited to 100 tokens per internal, but if the tokens had enough time to replenish to capacity, they would be able to go over the limit.

Example configuration:
```rust
RateLimitTokenBucketConfig::new(200, 100)
    .with_refill_interval(Duration::from_secs(1));
```

### Smoothening rate-limiting (avoiding spikes)

A common issue with rate-limiting implementations is that traffic tends to be very spiky. The tokens are refilled every time an interval passes, which means that under load, most of the requests will happen at the window edges.

> This can be solve using the token bucket algorithm by setting the configuration a bit differently. For example, if you want a rate-limit with 100 tokens that refills 100 tokens every 1 second, you can instead configure it to refill 1 token every 10 milliseconds. This still gives you a rate of 100 a second, but the refill happens more smoothly throughout the wanted one second internal.

Reviewer: ```suggestion
This can be solved using the token bucket algorithm by setting the configuration a bit differently. For example, if you want a rate-limit with 100 tokens that refills 100 tokens every 1 second, you can instead configure it to refill 1 token every 10 milliseconds. This still gives you a rate of 100 a second, but the refill happens more smoothly throughout the desired one second internal.
```
